# SeqPanther

SeqPanther tool consists set of commands as follows:

- **seqpatcher** : integrates sanger sequencing of missing regions of an incomplete assembly to the assembly. This command is a modification of [SeqPatcher](https://github.com/krisp-kwazulu-natal/seqPatcher) tool](https://github.com/krisp-kwazulu-natal/seqPatcher)

- **codon counter**: performs variant calling, generates nucleotide stats at variant sites and reports the impacts of nucleotide changes on amino acids in the translated proteins.

- **transform** : performs transformation of codon counter output in a format where the user can select a variant to integrate into the reference or assemblies using **nusub** command. The transformation can be performed after the removal of undesired changes in files generated by the codon counter or the first transformation is performed and undesired changes are removed from transformed data.

**nucsub** : integrates the alteration list generated by codoncounter in a given reference or assembled genome based on the user recommendations. <!--Providing options to users to select changes of their interests-->

## Operating system compatibility

Unix platforms.

## Dependencies

The tool relies on multiple external programs and python modules as listed below:

### External Tools

1.  Muscle

- To perform multiple sequence alignment
- Can be downloaded and installed from
- `conda install muscle` to install

2.  BLAT

- To query sequence location in the genome.
- `conda install blat` to install

3. MAFFT

- To align consensus sequence against the reference.
- `conda install mafft` to install

## Python modules

<!-- TODO: Versions need to be included -->

All the python module used for this program is listed in `requirements.txt`.

# Installation

## Manual installation

1.  `python setup.py install`

## Directly from the GitHub repo

`pip install git+https://github.com/codemeleon/seqPanther.git`

# Usages

The commands are self-explanatory.

seqPanther contains three commands, which can be accessed using `seqPanther --help`

## CodonCounter

This command can be used as `seqPanther CodonCounter`. The help is accessible at `seqPanther CodonCounter --help` or [CodonCounter GitHub page](https://github.com/codemeleon/CodonCounter)

## SeqIn

This command can be used as `seqPanther SeqIn`. The help is accessible at `seqPanther SeqIn --help` or [SeqIn GitHub page](https://github.com/codemeleon/SeqIn)

## SeqPatcher

This command can be used as `seqPanther SeqPatcher`. The help is accessible at `seqPanther` SeqPatcher --help` or [SeqPatcher GitHub page](https://github.com/CERI-KRISP/SeqPatcher)

## Example Dataset

An South African strain data are used as examples. Samples' short reads can be found at .... The reads were mapped to reference .... using ....

To shrink the size of the bam file, only spike protein-related areas were focused using the command `samtools view -h N47215.bam sars-cov-2:19000-26000 | samtools view -bS - > N47215_S_gene.bam`
GFF file for reference sequences was downloaded from [NCBI](https://www.ncbi.nlm.nih.gov/sars-cov-2/) and transformed to required need as [Trimmed gff](https://www.github.com/)
Reference sequence was also downloaded from [NCBI](https://www.ncbi.nlm.nih.gov/sars-cov-2/)

Consensus sequence was generated using command `samtools mpileup -uf N47215.bam | bcftools call -c --ploidy 1 | vcfutils.pl vcf2fq > SAMPLE_cns.fastq`

Fastq was converted to fasta using a custom command.

# Features

Codon Counter and SeqIn commands were specifically defined for this program. However, each sub-command can be installed and used independently of this program.

# Bug reporting

Please report bugs on this portal or associated GitHub repositories of the command

# Cite

If you use this tool, please cite it as

Bib for the opensource software journal (JOSS)
